### PR TITLE
Add second post on "Test Evaluation with ELK in the Pipeline"

### PR DESCRIPTION
**Pull request by dataduke**

### CODE REVIEW

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 1]
```diff
@@ -0,0 +1,290 @@
+---
+layout: post
+title: "Automated test evaluation with ELK in the pipeline: Implemented solution"
+date: "2016-02-16 06:16:02"
+icon: wrench
+categories: tech-stories
+authors: ["Benjamin N.", "Bastian K."]
+---
+
+We implemented a Selenium test report database with Elasticsearch, Logstash, Docker, CircleCI and Jenkins to ease the test evaluation process in our Continuous Delivery Pipeline. Last week we have already introduced you to the [background of the project](https://developer.epages.com/blog/2016/02/11/pipeline-elk-test-evaluation-1.html) and today we want to get your hands on the actual development steps. Hence, this post showcases the various parts of the implemented solution and discusses the pragmatic benefits for our pipeline and our speed-up for massive regression test evaluation.
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Last week we have already introduced you > Last week we already introduced you

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 2]
```diff
@@ -0,0 +1,290 @@
...
+
+Furthermore, this article should serve as an outline of the consolidated technical expertise gained throughout the engineering process of this project.
+
+## Implemented Solution
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Implemented Solution > Implemented solution

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+
+To get the big picture for splitting the Scrum epic into several stories with tasks and acceptance criteria we created a visualization, which could distinctly highlight the various parts that needed to be implemented. The first draft of the blueprint was sketched by hand and looked similar to this:
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: visualization > visualisation

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+{% image blog-pipeline-elk-test-evaluation-blueprint.png %} The blueprint of the solution architecture {% endimage %}
+
+As you can see above, several components of our infrastructure will be affected and also involved throughout the development of this project. The middle tier shows the essential interdigitation of the underlying job chain in our pipeline. Usually a CDP run involves several prepare jobs; then a huge amount of install and patch jobs are run in parallel on the various VMs of the vCenter (top tier); afterwards a fingerprint of all machines is created and finally the ESF testsuite (and others) are run onto all vCenter VMs. Sometimes the testsuite is even running against an ePages VM before, during and after patching has started (zero-down-time tests), so don't take the blueprint to literally.
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Usually a CDP >  Usually, a CDP

Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Center (top tier); afterwards a > Center (top tier). Afterwards a...

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 5]
```diff
@@ -0,0 +1,290 @@
...
+
+After the tests have been run, the JSON logs should have been created inside every single Jenkins job. As of now the tricky implementation of this project starts. We have decided to split the implementation in 5 parts and the next sections will explain each one step-by-step.
+
+### Part 1 - Define the Test Object and Extend the Test Suite Reporter
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Define the Test Object and Extend the Test Suite Reporter > Define the test object and extend the test suite reporter

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 6]
```diff
@@ -0,0 +1,290 @@
...
+
+All other fields cannot be derived from our test suite itself and therefore need to be enriched at the processing step in the pipeline. We will discuss these ingredients of the test object in the chapter about Logstash.
+
+### Part 2 - Set up Elasticsearch with Docker and CircleCI
+
+#### Dockerfile
+
+We decided to run the nodes of the [Elasticsearch](https://www.elastic.co/products/elasticsearch) cluster within effortlessly deployable Docker containers. To keep the entire setup at a reasonable level the reuse of the [official base image](https://hub.docker.com/_/elasticsearch/) was very helpful. In the `Dockerfile` we synced our timezone, prepared templating with [Jinja2](http://jinja.pocoo.org/docs/dev/) and installed two plugins for HTTP authorization and [administration](https://github.com/mobz/elasticsearch-head) via a web frontend that included a tabular document view and an extensive REST-console. We needed to create and use our own `docker-entrypoint` script as we wanted to map a few more Docker host directories than suggested by the official base image.
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: authorization > authorisation

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 7]
```diff
@@ -0,0 +1,290 @@
...
+
+#### Configuration
+
+Besides using variables in the configuration files of Elasticsearch the setup process was quite straightforward. We reduced complexity via a Bash script allowing to build the Docker image and start the container. The start script supports the setting of the needed environment variables for the configuration files and hands them over into the Docker container at runtime.
+For the daily operation of the Elasticsearch cluster we implemented a verbose mode in the start script as well as in the `docker-entrypoint` script so that we could monitor each step in the console output.
+
+#### Testing
+
+We versioned the entire source code on [GitHub](https://github.com/). The first file we added was the configuration file for the CircleCI job. The job basically clones the repository and tries to build and run the Docker container. After these described preparation steps several tests check if the Elasticsearch service is reachable from outside of the container and is working as expected. With this setup we could securely develop the `Dockerfile` and the Elasticsearch configuration files against the previously created tests.
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: You link just to the main GitHub page here. I assume that the repo with the code is not open-source?

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 8]
```diff
@@ -0,0 +1,290 @@
...
+
+#### Deployment
+
+Changes in the master branch trigger a Jenkins Job to run, which pulls the latest Docker image from the registry and restarts the container.
+
+### Part 3 - Set up Logstash with Docker and CircleCI
+
+#### Basics
+
+Within our architecture the Logstash container runs as a comprehensive log forwarder, which means it acts as processor, shipper and indexer of the test results. Hence, it's main purpose is to read, transform and feed the test objects as documents into our Elasticsearch cluster. For this purpose the configuration file defines the input and the output of Logstash as well as how the data has to be filtered before dispatching the events to Elasticsearch. The input is given by a file path, which can also contain a regular expression. As described above, we use a JSON log file that is parsed by Logstash. The filter is able to add and remove fields from the JSON object. The output declares where the formatted message should be sent to. In our case this is configurable by environment variables, which will be explained by the following paragraphs.
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: For this purpose the configuration file > For this purpose, the configuration file

Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: which will be explained by the following paragraphs. > which will be explained in the following paragraphs.

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 9]
```diff
@@ -0,0 +1,290 @@
...
+
+#### Configuration
+
+In the Logstash configuration, there is the possibility to use `if`-statements and environment variables. In addition to this, we decided to write our own templating engine based on the Jinja2 framework to achieve high flexibility in which variables can be feed into the rendering process. This allows us to have an environment specific configuration for each VM the Docker container is running on. To use this feature we forward some variables into our container. Our `docker-entrypoint` script renders the configuration templates and starts the Logstash agent.
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: In the Logstash configuration, there is the possibility to use > In the Logstash configuration it is possible to use

Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: variables can be feed into > variables can be fed into

Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: environment specific > environment-specific

Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: To use this feature we forward > To use this feature, we forward

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 10]
```diff
@@ -0,0 +1,290 @@
...
+{% endraw %}{% endhighlight %}
+
+
+The above code snippet shows how we push the output to our Elasticsearch cluster. The first line represents how we use our own templating engine. If the `if`-statement is `false`, the part configuring the output to Elasticsearch will be omitted. If it equals `true`, we use environment variables as well as information from our metadata for connection, document path and Elasticsearch index template, which contains index settings and the mappings of document fields to the desired data types. We also have written a similar construct to generate the output to stdout, an error log file and an info log file, which later contains successfully pushed data for monitoring reasons.
+
+#### Testing
+
+The CI part for our Logstash container is very similiar to that of the Elasticsearch container. The tests we run are different of course, but the process is the same.
+
+### Part 4 - Integrate Docker Containers in Pipeline with Jenkins
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Integrate Docker Containers in Pipeline with Jenkins >  Integrate Docker containers in pipeline with Jenkins

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 11]
```diff
@@ -0,0 +1,290 @@
...
+to-logstash-run-esf-tests-3829
+{% endhighlight %}
+
+All shipped test objects are saved to a Logstash info log, which is archived as a build artifact in the post-build section of Jenkins.
+
+#### Elasticsearch
+
+For our Elasticsearch Docker cluster we configured a new Jenkins job, which ensured that always the latest stable version of our image is used. We made sure to mount several host directories so that the Elasticsearch data, config and logs are stored on the VM. By firing up multiple Elasticsearch node containers joining the same cluster we achieved load-balance and shard redundancy.
+
+### Part 5 - Use the Elasticsearch Client to Evaluate the Test Results
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Use the Elasticsearch Client to Evaluate the Test Results > Use the Elasticsearch client to evaluate the test results

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 12]
```diff
@@ -0,0 +1,290 @@
...
+
+In the current state we use the [Elasticsearch Client](https://github.com/rdpatil4/ESClient) to monitor and analyse the test results. Here you can browse and filter the documents via dropdown menus for the index, which is our test object type (e.g. cdp-ui-tests) and the document type, which is the ePages repo id (e.g. 6.17.39). You can then narrow down the search with simple matches in the search field (e.g. only show tests with result FAILURE) or use the official [Lucence Query](http://www.lucenetutorial.com/lucene-query-syntax.html), which supports boolean operators, range matchers and more advanced features similar to a regex. It is possible to edit every single test object within the client by double-clicking a tabular row. Therefore, the `note` field can be used to add information about the error, like the cause of the error and the correspondong JIRA issue id.
+
+{% image blog-pipeline-elk-test-evaluation-client.png %} The test results in the Elasticsearch Client {% endimage %}
+
+Additionally, we also take advantage of three other ways to access our Elasticsearch cluster:
+
+* via the [Elasticsearch Head plugin](https://github.com/mobz/elasticsearch-head).
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: remove the fullstop at the end of the list items

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md [hunk 13]
```diff
@@ -0,0 +1,290 @@
...
+{% highlight text%}
+# URI Schema
+<protocol>://<domain>:<port>/<index>/<document_type>/_count?=<query_string>
+<protocol>://<domain>:<port>/<index>/<document_type>/_search?=<query_string>
+
+# Query String Example
+?pretty&size=1000&q=result:failure,skip AND epages_repo_id:*17.06.15
+{% endhighlight %}
+
+## Benefits and Conclusion
```
Comment (BirgitBader) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Benefits and Conclusion > Benefits and conclusion

**Comment by BirgitBader** [comment]
@dataduke and @bklein1 This is such a great article. We definitely like to see more of that.

**Comment by BirgitBader** [comment]
Add the very end of the post please add:

## Related post

[Automated test evaluation with ELK in the pipeline: Background story](https://developer.epages.com/blog/2016/02/11/pipeline-elk-test-evaluation-1.html)

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md (hunk 7 above)
Comment (dataduke) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: Exactly, it's private ePages-de repo. The link to the landing page might be misleading here. I will remove it.

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md (hunk 2 above)
Comment (dataduke) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: I will adapt the headline capitalization to fit our policy :+1: 

Some popular newspapers/websites (e.g. NYT, WSJ) capitalize nearly all words except some filling words ("the,to,in"). So it was a tough choice. Should have look it up in your space first. :-)

File: _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md (hunk 9 above)
Comment (dataduke) [inline _posts/2016/2016-02-16-pipeline-elk-test-evaluation-2.md]: fed ... missed the past participle in the passive tense. :-(

**Comment by dataduke** [comment]
@BirgitBader I think we definitely need to change the title a bit. 

As I was observing my twitter account today I realized that Bastian and myself were a bit too much gathered into our tech stuff ("betriebsblind") and we did not realize that "ELK or pipeline" are such ambiguous terms and their meaning outside the DevOps bubble is nearly unknown.

For example if you google "ELK" you can click through hundreds of results until you get a single meaningful match with our tech context. The same is true for "test" or "pipeline".

I am no marketing or SEO expert, but this is definitely not good at all. Even on twitter hardly anybody besides some epages employees could find the tweet.

Hm, so long story short. I propose:

"Selenium test evaluation with Docker and Elasticsearch: Background story"
"Selenium test evaluation with Docker and Elasticsearch: Implemented Solution"

Can we discuss this briefly tomorrow? ... I think the keywords are important.

PS: The file names should stay the same, so that all tweeted links are still valid.

**Comment by BirgitBader** [comment]
Call me, when you're in the office - we can fix this of course.

**Comment by dataduke** [comment]
superseeded by https://github.com/ePages-de/epages-docs/pull/381

**Pull request closed without merging**